test: add --test-files-glob flag

p-mcgowan opened the pull request:
test: add --test-files-glob flag

Overrides the default `kDefaultPattern` for test file globs to allow for
default file pattern matching, while still allowing other options or
single, targeted file names for testing.

This also might resolve issues with the defaults when using strip-types
since the defaults can, currently, only be overridden using the positional
`arguments` (see #56546).

Fixes: #51384
Ref: #56546

I believe this is a notable-change, I'll leave that up the reviewers though.

I also [considered modifying the arg parser](https://github.com/nodejs/node/issues/51384#issuecomment-2994199816) when the `--test` flag is present to continue parsing options that come after positionals.
I think it's still (maybe) a decent solution (even if it behaves differently than other commands) since the test positionals are globs instead of files (like most other commands).

Happy to cleanup and push that up alternatively if that's preferred, or try something else if neither work - would like to resolve the current issues so that the default runner is a drop-in replacement for everything else.

Also note: still getting https://github.com/nodejs/node/issues/44003 locally, I believe the solution is to enable another interface but I didn't dig too far into it, admittedly.

```
Failed tests:
out/Release/node --expose-internals --test-reporter=./test/common/test-error-reporter.js --test-reporter-destination=stdout /home/patrickmcgowan/source/p-mcgowan/node/test/parallel/test-http2-invalid-last-stream-id.js
out/Release/node --expose-internals --test-reporter=./test/common/test-error-reporter.js --test-reporter-destination=stdout /home/patrickmcgowan/source/p-mcgowan/node/test/parallel/test-http2-premature-close.js
out/Release/node --test-reporter=./test/common/test-error-reporter.js --test-reporter-destination=stdout /home/patrickmcgowan/source/p-mcgowan/node/test/parallel/test-net-socket-connect-without-cb.js
out/Release/node --expose-internals --test-reporter=./test/common/test-error-reporter.js --test-reporter-destination=stdout /home/patrickmcgowan/source/p-mcgowan/node/test/parallel/test-tcp-wrap-listen.js
make[1]: *** [Makefile:315: jstest] Error 1
make: *** [Makefile:342: test] Error 2
```

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

pmarchini commented on README.md:
```diff
@@ -409,6 +409,8 @@ For information about the governance of the Node.js project, see
   **Moshe Atlow** <<moshe@atlow.co.il>> (he/him)
 * [MrJithil](https://github.com/MrJithil) -
   **Jithil P Ponnan** <<jithil@outlook.com>> (he/him)
+* [p-mcgowan](https://github.com/p-mcgowan) -
```
Hey @p-mcgowan, thanks for opening this PR, I'll take a look ASAP!

Only one important note: the term collaborator has a specific meaning and is not the same as "contributor".
Please remove this change from the commit 😊
You can find more details about this point here: https://github.com/nodejs/node/blob/main/GOVERNANCE.md#collaborators

p-mcgowan replied:
Well that's embarrassing, thanks - I've removed it.

marco-ippolito commented:
Thanks for the PR can you add a empty line at the of the fixtures?

p-mcgowan commented:
Yeah I was wondering about that - my editor is using the top-level [`.editorconfig`](https://github.com/nodejs/node/blob/main/.editorconfig#L26) which has:
```ini
[{test/fixtures,deps,tools/eslint/node_modules,tools/gyp,tools/icu,tools/msvs}/**]
insert_final_newline = false
```

Should I also remove the js fixtures from that glob? Or just manually add the newline and leave .editorconfig alone?

p-mcgowan committed:
test: add --test-files-glob flag

Overrides the default `kDefaultPattern` for test file globs to allow for
default file pattern matching, while still allowing other options or
single, targeted file names for testing.

This also might resolve issues with the defaults when using strip-types
since the defaults can, currently, only be overridden using the
positional `arguments` (see #56546).

Fixes: https://github.com/nodejs/node/issues/51384
Ref: https://github.com/nodejs/node/issues/56546

p-mcgowan commented:
I've just added the newlines manually - I will say it's out of scope to fix the editorconfig / fixture files (that could be a lit-fix for someone else I think).

I've also fixed the js lint and commit message failing runs

p-mcgowan commented:
> How is different from users passing the positional glob argument?

In short: it's an option instead of positional, which sounds a bit stupid but actually solves some problems.

I mentioned that it seems duplicated in the [comment on #51384](https://github.com/nodejs/node/issues/51384#issuecomment-2997698260) and suggested [alternatives in another related ticket](https://github.com/nodejs/node/issues/56546#issuecomment-2705731169), but ultimately it provides a solution (even if only until a better one comes along) to 2 things:
 - it allows for people to set "positional globs" in the options, which then allow for more options afterwards or globs (solving #51384)
 - it also provides a solution to some issues where the default glob patterns match files that shouldn't be matched (eg .ts files) that is unexpected.
 
So it largely serves the same purpose, but as an option instead of positional it means you can do something like default the glob to `dist/**/*.spec.js` and then pass in `--test-only` afterwards.

I'm open to other possible solutions, I just wanted to get something in and working but I think this also brings it more inline to how people are familiar with using test runners (mocha, vite, deno, etc). Or rather, the original proposed solution on #51384 that I created [a different branch](https://github.com/nodejs/node/compare/main...p-mcgowan:node:feat/node-test-flags-can-come-after-positional) for was closer to that (simply continuing to parse cli flags after seeing positionals) but I felt that was too disruptive a change, but I'd honestly prefer that option.

p-mcgowan commented:
Actually fixed the commit message this time, and ran core-validate-commit to tripple check.

p-mcgowan commented:
Actually to expand on the above reasoning:

The original idea was to continue to parse node flags after positionals when `--test` was set, but this introduced 2 new issues:
1.
it meant that the cli handles arguments differently when using `--test`, which was an inconsistency that felt kind of wrong.
2.
it meant that even though you could provide a default glob now and options after, single files would be ignored.

For example, with an npm script `"test": "NODE_ENV=test node --test --experimental-strip-types 'test/**/*.spec.ts'"` (or any other default flags you might be using), if you ran `npm test -- --test-only` it would work, but if you ran `npm test some/single/file.ts` it would execute all tests, and the one provided.

So the core issue is not that positionals have to come after flags, but that there is no other way to override the default glob other than positionals. So by providing that as an option, the npm script would then be `"test": "NODE_ENV=test node --test --experimental-strip-types --test-files-glob='test/**/*.spec.ts'"`, and `npm test -- --test-only` works as well as `npm test some/single/file.ts`.

p-mcgowan commented:
Not sure why codecov is reporting that check as uncovered, I wrote tests specifically for it. Grep'd the logs, looks like all possible branches are either already covered, or covered by one of the 2 new tests.

![image](https://github.com/user-attachments/assets/a6dacd30-e6e6-4c92-9e2d-6e971c5aacb5)

marco-ippolito commented on src/node_options.cc:
```diff
@@ -828,6 +828,14 @@ EnvironmentOptionsParser::EnvironmentOptionsParser() {
             &EnvironmentOptions::test_name_pattern,
             kAllowedInEnvvar,
             OptionNamespaces::kTestRunnerNamespace);
+  AddOption("--test-files-glob",
+            "set the default glob pattern for matching test files (default: "
+            "'**/{test,test/**/*,test-*,*[._-]test}.{<extensions>}' where "
+            "<extensions> is 'js,mjs,cjs' or 'js,mjs,cjs,ts,mts,cts' when using"
+            " --experimental-strip-types)",
+            &EnvironmentOptions::test_files_glob,
+            kAllowedInEnvvar,
+            OptionNamespaces::kTestRunnerNamespace);
```
```suggestion
  AddOption("--test-files-glob",
            "set the default glob pattern for matching test files",
            &EnvironmentOptions::test_files_glob,
            kAllowedInEnvvar,
            OptionNamespaces::kTestRunnerNamespace);
```

marco-ippolito commented:
The man needs to be updated too https://github.com/nodejs/node/blob/main/doc/node.1

p-mcgowan commented:
Awesome, thanks. I don't think I knew node had a manpage...

p-mcgowan committed:
enh: cleanup tests

remove conosle logs, organize covereage into subtests, try to fool
codecov into realizing the tests are fine.

p-mcgowan commented:
Reports of my inactivity are greatly exaggerated. I'm watching you, github bot... 

Not sure if this will make it in, seems like there is activity / discussion elsewhere around the test runner behaviour currently.

If not - feel free to close it. In the mean time (at least until npm drops support for it as it's deprecated) I've just been working around it with:

```json
"scripts": {
  "test": "NODE_ENV=test node --test ${npm_config_test:-\"dist/test/**/*.spec.js\"}"
}
```
and `npm test` or `npm test --test=-- dist/test/single-file.js`
